### DISCUSSION TITLE
Weird bug on vercel deployment

### DISCUSSION THREAD

**Discussion by saile515** [vercel/next.js#36158] [Help]
I have a contact form on my site, that sends and email via an api function. Everything works great on local, but on production (only on vercel) I get this strange error on the function logs:

```
2022-04-14T14:08:08.619Z	95566025-da05-4d4a-be14-753b49e92efc	ERROR	400
RequestId: 95566025-da05-4d4a-be14-753b49e92efc Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

This doesn't describe the issue at all, and I can't find any information about what it is. The api function doesn't seem to run at all, since I have console logs in the handler, which doesn't log.

I am lost on this and would really appreciate some help, thanks.

**Comment by khuezy** [comment]
What does your handler look like? Can you try wrapping it in `try/catch` and log the catch?

**Reply by saile515** [reply]
This is the handler function:
```ts
export default function handler(req: NextApiRequest, res: NextApiResponse) {
	try {
		const data = req.body as Data;
		const mailOptions = {
			from: process.env.EMAIL,
			to: process.env.EMAIL,
			subject: `Potentiell kund - ${data.name}`,
			text: `Email: ${data.email}\nNamn: ${data.name}${data.tel ? "\nTelefon: " + data.tel : ""}${data.companyName ? "\nFöretagsnamn: " + data.companyName : ""}\n\nIdé:\n${data.idea}`,
		};

		return new Promise((resolve, reject) => {
			transporter.sendMail(mailOptions, (error, info) => {
				if (error) {
					res.redirect("/");
					res.end();
					reject(400);
				} else {
					res.redirect("/");
					res.end();
					resolve(200);
				}
			});
		});
	} catch (err) {
		console.log(err);
	}
}
```

The try catch doesn't log anything in the vercel function logs, only the error appears.

**Comment by icyJoseph** [comment]
A couple of follow up questions based on your answer:

- Where does transporter come from?
- What's the size of data (req.body)?
- Is `req.body` properly formed? AKA correct JSON?
- You say its only on Vercel, but perhaps some credentials are missing to transport the email?

If I remember correctly, this is totally busting the runtime of the lambda, hence why, not even the try catch can handle the problem. In other words, the entire node runtime falls apart. And, if I remember correctly, these are AWS lambdas that Vercel uses, which makes me thing that, either, the lambda is taking too long, or it is consuming too much memory, or something even worse.

**Reply by saile515** [reply]
The data is only a few kilobytes at most. The transporter is outside of the handler, which I realized was a mistake, but it didn't work when I moved it into the handler either.

I assume the body is correctly formatted, as it works locally. I also think it would throw a regular error if the credentials were wrong, as it did that locally when testing there (before I fixed that issue)

**Reply by icyJoseph** [reply]
Right, it is ok if its outside the handler, but then we must make sure that the runtime can interpret the code outside the lambda, when it does its spin up.

For argument's sake take the following:

```ts
throw new Error("boom");

const handler = () => {};

export default handler;
```

Though the handler is fine, there's an error at the top of the module, which we cannot control. 

Try something like:

```ts
let transporter: TransporterType;

try {
transporter = new Transporter(); // or w.e.
} catch (err) {
 console.log(err)
}

// yada yada, the rest of the function and it's parameters
export default function handler() {
}
```

Not that this should fix the error though.

What node version are you using? Vercel supports only 12.x and 14.x, perhaps some of the code you are using (or a dependency) uses a feature not yet supported.

We'll get to the bottom of this!

**Reply by saile515** [reply]
Moving the transporter initialization into a try catch gave the same result. I am using node v16.13.0, These are my dependencies:
```json
"dependencies": {
    "dotenv": "^16.0.0",
    "next": "12.1.4",
    "nodemailer": "^6.7.3",
    "react": "18.0.0",
    "react-dom": "18.0.0"
  },
  ```
